### PR TITLE
Add UniG2U benchmark task with model support

### PR DESCRIPTION
**Pull request by nssmd**
## Summary

Add UniG2U (Unified Generation-to-Understanding) benchmark with 11 sub-tasks and 5 model implementations.

- Two evaluation modes: `unig2u` (standard understanding) and `unig2u_GtA` (two-stage Visual CoT)
- GtA triggered explicitly via `generation_kwargs.visual_cot: true`
- LLM-judge supports Azure CLI endpoint (`AZURE_OPENAI_ENDPOINT`), TRAPI, and OpenAI backends
- New models: Ovis-U1, ILLUME+, MMaDa, Qwen-Image-Edit, Bagel-UniG2U (with interleaved generation)

## Changes

**Modified:**
- `lmms_eval/models/__init__.py` — register 5 new models
- `lmms_eval/models/simple/ovis_u1.py` — fix dtype for Stage1 generation
- `lmms_eval/models/simple/qwen_image_edit.py` — fix device_map for diffusion pipeline
- `.gitignore` — exclude Bagel/MMaDA repos

**New model files:**
- `lmms_eval/models/simple/ovis_u1.py` (understanding + GtA)
- `lmms_eval/models/simple/illume_plus.py` (understanding + visual CoT)
- `lmms_eval/models/simple/mmada.py` (understanding only)
- `lmms_eval/models/simple/qwen_image_edit.py` (understanding + GtA via editing)
- `lmms_eval/models/simple/bagel_unig2u.py` (understanding + GtA + interleaved generation)

**New task directory:**
- `lmms_eval/tasks/unig2u/` — 83 files (yaml configs, utils, README)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -81,3 +81,8 @@ CLAUDE.md
 .opencode
 .ignored/
 .worktrees/
+Bagel/
+MMaDA/
+Bagel/
+MMaDA/
```
Comment (kcz358) [inline .gitignore]: Duplicate here.

**Review by kcz358 — COMMENTED** [review]
I'm fine with most of the implementation. If we're not able to have the author revise the README to support loading via load_dataset using the dataset name and split, then I'm okay with the current structure as is.

Given the number of tasks involved, it might be helpful to have the agent organize the unig2u folder a bit more clearly so that each task appears more structured.

File: .gitignore (hunk 1 above)
Comment (nssmd) [reply .gitignore]: Fixed, thanks！

File: lmms_eval/tasks/unig2u/utils.py [hunk 2]
```diff
@@ -0,0 +1,2891 @@
...
+            )
+
+    def chat_completion(self, *, messages, **kwargs) -> str:
+        resp = self.client.chat.completions.create(
+            model=self.deployment, messages=messages, **kwargs
+        )
+        return resp.choices[0].message.content
+
+
+_JUDGE_CLIENT: _JudgeClient | None = None
```
Comment (mwxely) [inline lmms_eval/tasks/unig2u/utils.py]: `_JudgeClient | None = None` requires Python 3.10+. The union type operator `X | Y` is not available in 3.9. Please use `Optional[_JudgeClient]` from `typing` instead, consistent with the rest of the codebase.

File: lmms_eval/tasks/unig2u/utils.py [hunk 3]
```diff
@@ -0,0 +1,2891 @@
...
+    )
+
+    return f"[GEN_PROMPT]{CHARTQA_GEN_PROMPT}[/GEN_PROMPT][QUESTION]{question_with_aux}[/QUESTION]"
+
+
+# ======================================================================
+# mmsi_bench
+# ======================================================================
+
+eval_logger = logging.getLogger("lmms-eval")
```
Comment (mwxely) [inline lmms_eval/tasks/unig2u/utils.py]: `eval_logger` is reassigned here from the loguru logger (imported at line 22) to `logging.getLogger("lmms-eval")`. This means functions defined above this line use loguru, while functions below use stdlib logging, creating inconsistent log output. The same reassignment happens again at line 2455. Please pick one logging approach and use it consistently throughout the file.

File: lmms_eval/models/__init__.py [hunk 4]
```diff
@@ -22,7 +22,8 @@
     "glm4v": "GLM4V",
     "auroracap": "AuroraCap",
     "bagel": "Bagel",
-    "bagel_umm": "BagelUMM",
+    "bagel_umm": "lmms_eval.models.simple.bagel.BagelUMM",
```
Comment (mwxely) [inline lmms_eval/models/__init__.py]: The value for `bagel_umm` was changed from a short class name (like all other entries) to a fully-qualified path `lmms_eval.models.simple.bagel.BagelUMM`. This is inconsistent with the rest of the dict. Was this intentional? If the old registration was broken, please document why.

File: lmms_eval/models/simple/illume_plus.py [hunk 5]
```diff
@@ -0,0 +1,2167 @@
...
+                        eval_logger.warning("Retrying with attn_implementation='eager'")
+                        model_kwargs["attn_implementation"] = "eager"
+                        self._model = AutoModel.from_pretrained(pretrained, **model_kwargs)
+                else:
+                    raise
+
+            self._model = self._model.eval()
+            self._config = self._model.config
+            
+            # MEMORY OPTIMIZATION: Enable gradient checkpointing for inference
```
Comment (mwxely) [inline lmms_eval/models/simple/illume_plus.py]: `gradient_checkpointing_enable()` is designed for training (recomputing activations during backward pass). During inference with `torch.no_grad()`, it provides no memory benefit and adds computational overhead. Please remove this call.

File: lmms_eval/models/simple/qwen_image_edit.py [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+        if self.mode != "editing":
+            raise ValueError(
+                "Uni-MMMU interleaved mode requires editing mode. "
+                "Use: --model_args mode=editing"
+            )
+        
+        # We need understanding model for text generation
+        # Load it temporarily if not already loaded
+        if not hasattr(self, '_understanding_model'):
+            eval_logger.info("Loading understanding model for text generation...")
```
Comment (mwxely) [inline lmms_eval/models/simple/qwen_image_edit.py:309]: Loading a second full `QwenImageEdit` instance inside `generate_uni_mmmu_interleaved` will duplicate the entire model in GPU memory. For large models, this will cause OOM on most hardware. Consider sharing the model instance or using a lighter-weight approach for the understanding path.

File: lmms_eval/models/simple/mmada.py [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+            generated_ids = output_ids[:, input_ids.shape[1] :]
+            response_text = self.uni_prompting.text_tokenizer.batch_decode(
+                generated_ids,
+                skip_special_tokens=True,
+                clean_up_tokenization_spaces=False,
+            )[0]
+
+            return response_text
+
+        except Exception as e:
```
Comment (mwxely) [inline lmms_eval/models/simple/mmada.py:320]: Errors are silently caught and an empty string is returned. This means evaluation will count model crashes as wrong answers without any way to distinguish "model got it wrong" from "model errored". Consider tracking error counts and reporting them in the final results.

**Review by mwxely — DISMISSED** [review]
## Code Review: Add UniG2U benchmark task with model support

Thanks for the comprehensive contribution! The unified G2U benchmark with visual CoT support is a valuable addition. Below are some suggestions for improvement.

### High Priority: Function Name Collisions in utils.py

The 2891-line `utils.py` defines `process_results`, `doc_to_visual`, `doc_to_text`, and `aggregate_results` **multiple times** for different sub-tasks. Later definitions silently overwrite earlier ones, so YAML configs referencing `utils.process_results` will dispatch to the last-defined version (realunify), not the intended sub-task version. This will produce incorrect evaluation results for some sub-tasks.

**Suggestion:** Split into per-task utility modules or use unique prefixed names. (This aligns with @kcz358's suggestion to organize the unig2u folder more clearly.)

### Other Suggestions
1. **Python 3.10+ syntax** (`_JudgeClient | None` at L148) — use `Optional[_JudgeClient]` for 3.9 compatibility
2. **`eval_logger` reassigned** from loguru to stdlib logging mid-file (L261, L2455)
3. **`parse_response` random fallback** (L2567) inflates accuracy by ~25% on unparseable responses — consider returning None and scoring as 0
4. **`gradient_checkpointing_enable()`** called during eval in illume_plus.py L402 (no benefit during inference)
5. **Dual model loading** in qwen_image_edit.py L310 may cause OOM
6. **`bagel_umm` registration** changed to fully-qualified path in `__init__.py` L25, inconsistent with other entries

**Review by mwxely — COMMENTED** [review]
## Code Review: Add UniG2U benchmark task with model support

Thanks for the comprehensive contribution! The unified G2U benchmark with visual CoT support is a valuable addition. Below are some suggestions for improvement.

### High Priority: Function Name Collisions in utils.py

The 2891-line `utils.py` defines `process_results`, `doc_to_visual`, `doc_to_text`, and `aggregate_results` **multiple times** for different sub-tasks. Later definitions silently overwrite earlier ones, so YAML configs referencing `utils.process_results` will dispatch to the last-defined version (realunify), not the intended sub-task version. This will produce incorrect evaluation results for some sub-tasks.

**Suggestion:** Split into per-task utility modules or use unique prefixed names. (This aligns with @kcz358's suggestion to organize the unig2u folder more clearly.)

### Other Suggestions
1. **Python 3.10+ syntax** (`_JudgeClient | None` at L148) — use `Optional[_JudgeClient]` for 3.9 compatibility
2. **`eval_logger` reassigned** from loguru to stdlib logging mid-file (L261, L2455)
3. **`parse_response` random fallback** (L2567) inflates accuracy by ~25% on unparseable responses — consider returning None and scoring as 0
4. **`gradient_checkpointing_enable()`** called during eval in illume_plus.py L402 (no benefit during inference)
5. **Dual model loading** in qwen_image_edit.py L310 may cause OOM
6. **`bagel_umm` registration** changed to fully-qualified path in `__init__.py` L25, inconsistent with other entries
7. **`generate_visual_cot` output type** — would a `generation_kwargs` flag be simpler since most models just delegate to `generate_until`?

**Review by mwxely — APPROVED** [review]
LGTM, thanks for addressing the feedback!
